### ISSUE TITLE
a question about making NAT decision using bpf + netfilter data structure

### ISSUE THREAD

**Issue by walidbarakat** [ti-mo/conntracct#37]
Hi,
i'm not sure if this the right place or the right way to ask this question, i'm sorry in advance if it's not.

i find this project is inspiring for a private project i'm working on, a NAT server in pure XDP. but to have a robust nat functionality i wanted to track the connection (storing and fetching tracking data) using the already exist netfilter conntrack data structure and functions to help with packet processing in XDP, i have some knowledge about nf_conntrack but i feel i miss something, i'm not sure exactly if i can just recieve the packet and check with netfilter conntrack if it's a new connection or not, for example if it's a new connection i can just call a netfiler conntrack function which can make decision about what port i can assign as public source port and another info to a tuple and when the reply back i can check in nf conntrack that it's a reply and just update the packet info and pass to the internal network. 

i feel confused a little but if i managed to explain what i'm thinking about and you have a hint or a tip to direct me, i'd be gratful.

thanks.

**Comment by ti-mo** [comment]
Hey @walidbarakat. This is indeed not the ideal platform for this type of questions. As previously discussed on Cilium Slack:

- Calling arbitrary kernel functions from eBPF is not (yet) supported
- XDP hooks run before iproute2 processing takes place, so the plan you had in mind wouldn't quite work
- Using existing nf_conntrack structs to build an XDP-based conntrack around wouldn't be ideal, since they contain many fields related to kernel plumbing (e.g. spinlocks and other synchronization primitives) and a web of ifdefs based on kernel config.

In short, you're better off rolling your own, purpose-built datapath. :slightly_smiling_face:
